File: AnYang/String/38. Count and Say.py
# 2019-07-06 Sat

# 看了半天才读懂题
class Solution:
    def countAndSay(self,n):
        if n == 1:
            return '1'
        s = "1"
        for j in range(1, n):
            t = s[0]
            temp = ""
            count = 0
            for i in range(len(s)):
                if(t==s[i]):
                    count += 1
                else:
                    temp = temp+str(count)+t
                    t = s[i]
                    count = 1
            s = temp+str(count)+t
        return s

File: tobehe/38.count-and-say.cpp
/*
 * @lc app=leetcode id=38 lang=cpp
 *
 * [38] Count and Say
 */
class Solution {
public:
  string countAndSay(int n) 
    {
        string r("1");
        for(int i = 2; i <= n;i++)
        {
            r = produce(r);    
        }
        return r;
    }

    string produce(string r)
    {
        string ans;
        int length = r.length();
        int i = 0;
        for(i = 0; i < length; )
        {
            char count = '0';
			int num = 0;
            while( num < length - i && r[i] == r[i + num] )
            {
                count++;
				num++;
            }
            
            ans += count;
            ans += r[i];
            i+= num;
        }
        return ans;
    }

};


